src/hotspot/cpu/x86/x86.ad

```diff
@@ -7866,121 +7865,110 @@ instruct loadMask(legVec dst, legVec src) %{
-      assert(UseAVX >= 2, "required");
+      assert(UseAVX > 0, "required");
       int src_vlen_enc = vector_length_encoding(this, $src);
       __ vpabsb($dst$$XMMRegister, $src$$XMMRegister, src_vlen_enc);
     }
   %}
   ins_pipe( pipe_slow );
 %}
 
-instruct storeMask2B(vec dst, vec src, immI_2 size) %{
-  predicate(Matcher::vector_length(n) <= 8 &&
-            n->in(1)->bottom_type()->isa_vectmask() == NULL);
-  match(Set dst (VectorStoreMask src size));
-  format %{ "vector_store_mask $dst,$src\n\t" %}
-  ins_encode %{
-    assert(UseSSE >= 3, "required");
-    __ pabsw($dst$$XMMRegister, $src$$XMMRegister);
-    __ packsswb($dst$$XMMRegister, $dst$$XMMRegister);
-  %}
-  ins_pipe( pipe_slow );
-%}
-
 instruct vstoreMask2B(vec dst, vec src, immI_2 size) %{
-  predicate(Matcher::vector_length(n) == 16 && !VM_Version::supports_avx512bw() &&
-            n->in(1)->bottom_type()->isa_vectmask() == NULL);
+  predicate(Matcher::vector_length(n) <= 16 && n->in(1)->bottom_type()->isa_vectmask() == NULL);
   match(Set dst (VectorStoreMask src size));
-  effect(TEMP dst);
-  format %{ "vector_store_mask $dst,$src\t!" %}
+  effect(TEMP_DEF dst);
+  format %{ "vector_store_mask $dst, $src \t! elem size is $size byte[s]" %}
   ins_encode %{
     int vlen_enc = Assembler::AVX_128bit;
-    __ vextracti128($dst$$XMMRegister, $src$$XMMRegister, 0x1);
-    __ vpacksswb($dst$$XMMRegister, $src$$XMMRegister, $dst$$XMMRegister,vlen_enc);
-    __ vpabsb($dst$$XMMRegister, $dst$$XMMRegister, vlen_enc);
-  %}
-  ins_pipe( pipe_slow );
-%}
-
-instruct storeMask4B(vec dst, vec src, immI_4 size) %{
-  predicate(Matcher::vector_length(n) <= 4 && UseAVX <= 2);
-  match(Set dst (VectorStoreMask src size));
-  format %{ "vector_store_mask $dst,$src\t!" %}
-  ins_encode %{
-    assert(UseSSE >= 3, "required");
-    __ pabsd($dst$$XMMRegister, $src$$XMMRegister);
-    __ packssdw($dst$$XMMRegister, $dst$$XMMRegister);
-    __ packsswb($dst$$XMMRegister, $dst$$XMMRegister);
+    int vlen = Matcher::vector_length(this);
+    if (vlen <= 8 && UseAVX <= 2) {
+      assert(UseSSE >= 3, "required");
+      __ pabsw($dst$$XMMRegister, $src$$XMMRegister);
+      __ packuswb($dst$$XMMRegister, $dst$$XMMRegister);
+    } else {
+      assert(UseAVX > 0, "required");
+      __ vextracti128($dst$$XMMRegister, $src$$XMMRegister, 0x1);
+      __ vpacksswb($dst$$XMMRegister, $src$$XMMRegister, $dst$$XMMRegister, vlen_enc);
+      __ vpabsb($dst$$XMMRegister, $dst$$XMMRegister, vlen_enc);
+    }
   %}
   ins_pipe( pipe_slow );
 %}
 
 instruct vstoreMask4B(vec dst, vec src, immI_4 size) %{
-  predicate(Matcher::vector_length(n) == 8 && UseAVX <= 2);
+  predicate(UseAVX <= 2 && Matcher::vector_length(n) <= 8 && n->in(1)->bottom_type()->isa_vectmask() == NULL);
   match(Set dst (VectorStoreMask src size));
-  format %{ "vector_store_mask $dst,$src\t!" %}
-  effect(TEMP dst);
+  format %{ "vector_store_mask $dst, $src \t! elem size is $size byte[s]" %}
+  effect(TEMP_DEF dst);
```

Comment:
The original storeMask4B didn't need TEMP dst and so was a separate rule.

Reply:
TEMP_DST ensures creation of separate machine operands which gets allocated same register as DST, thus preventing any accidental sharing of register with any other temporary machine operand over the node.  In this case since there is no other explicit temporary register used to effectively, TEMP_DEF  is same at DEF for all practical purpose.